### PR TITLE
feat(server): Support CLIENT TRACKING subcommand (1/2)

### PR DESCRIPTION
**Pull request by theyueli**
fixes https://github.com/dragonflydb/dragonfly/issues/2139

Implement CLIENT TRACKING subcommand.
 
Track the keys of a readonly command by maintaining mapping that maps keys to the sets of tracking clients.

Also include a transaction interface `Refurbish()`  that allows the reuse of a transaction, which is used for tracking the keys involved in a readonly command.

### CODE REVIEW

File: src/server/main_service.cc [hunk 1]
```diff
@@ -1465,6 +1485,9 @@ void Service::Quit(CmdArgList args, ConnectionContext* cntx) {
     cntx->SendOk();
   using facade::SinkReplyBuilder;
 
+  // turn off tracking for this client
```
Comment (dranikpg) [inline src/server/main_service.cc:1488]: No need for it here,.this path calls OnClose

File: src/server/server_family.cc [hunk 2]
```diff
@@ -1012,6 +1012,7 @@ string ServerFamily::GetReplicaMasterId() const {
 
 void ServerFamily::OnClose(ConnectionContext* cntx) {
   dfly_cmd_->OnClose(cntx);
+  cntx->conn()->SetClientTrackingSwitch(false);
```
Comment (dranikpg) [inline src/server/server_family.cc:1015]: I assume we had `MainService:: OnClose` where watch, pubsub and monitoring are cleaning up. Let's add it there so all of them are in one place

File: src/server/main_service.cc [hunk 3]
```diff
@@ -1147,6 +1154,19 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
...
+  // if this is a read command, and client tracking has enabled,
+  // start tracking all the updates to the keys in this read command
+  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn()) {
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      auto keys = t->GetShardArgs(shard->shard_id());
+      vector<string_view> keys_to_track{keys.begin(), keys.end()};
+      return OpTrackKeys(t->GetOpArgs(shard), dfly_cntx, keys_to_track);
+    };
+
+    dfly_cntx->transaction->Refurbish();
```
Comment (dranikpg) [inline src/server/main_service.cc:1166]: 1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)
2. DispatchCommand is not used for all commands, but we can ignore that for now
3. Still reminding that we can track keys when the transaction concludes and no "hop" is needed solely for this, but I also assume it's not a must for the first version

File: src/server/main_service.cc [hunk 4]
```diff
@@ -1147,6 +1154,19 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
     dfly_cntx->reply_builder()->CloseConnection();
   }
 
+  // if this is a read command, and client tracking has enabled,
+  // start tracking all the updates to the keys in this read command
+  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn()) {
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      auto keys = t->GetShardArgs(shard->shard_id());
+      vector<string_view> keys_to_track{keys.begin(), keys.end()};
```
Comment (dranikpg) [inline src/server/main_service.cc:1162]: Why copy it into a vector and not use the span?

File: src/server/db_slice.cc [hunk 5]
```diff
@@ -1338,4 +1338,27 @@ void DbSlice::ResetUpdateEvents() {
...
+  for (auto key : keys) {
+    if (client_tracking_map_.find(key) == client_tracking_map_.end()) {
+      DVLOG(2) << "The key " << key << " has not been tracked by any client.";
+      DVLOG(2) << "Creating a new entry in the tracking table for this key.";
+      absl::flat_hash_set<facade::Connection::WeakRef, Hash> tracking_client_set{conn};
+      client_tracking_map_[key] = tracking_client_set;
+    } else {
+      DVLOG(2) << "The key " << key << " exists in the client tracking table.";
+      DVLOG(2) << "Inserting client ID " << conn.GetClientId() << " into its tracking client set: ";
+      client_tracking_map_[key].insert(conn);
```
Comment (dranikpg) [inline src/server/db_slice.cc:1359]: I know you have different output for now, but at the end you don't need this if  

map[key].insert() works even if key is not in map

File: src/facade/dragonfly_connection.h [hunk 6]
```diff
@@ -167,7 +167,7 @@ class Connection : public util::Connection {
     bool EnsureMemoryBudget() const;
 
     bool operator<(const WeakRef& other);
-    bool operator==(const WeakRef& other);
+    bool operator==(const WeakRef& other) const;
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.h:170]: Yes forgot this 😅

File: src/server/main_service.cc (hunk 3 above)
Comment (theyueli) [reply src/server/main_service.cc:1166]: > 1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)

I'm wondering, the transaction probably has already been used before the tracking (e.g. in the previous command invoking?) and I need to do a refurbish here before doing my own singlehop?

Comment (theyueli) [reply src/server/main_service.cc:1166]: > 1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)
> 2. DispatchCommand is not used for all commands, but we can ignore that for now
> 3. Still reminding that we can track keys when the transaction concludes and no "hop" is needed solely for this, but I also assume it's not a must for the first version

Sounds good to me, I will look into this later.

File: src/server/main_service.cc (hunk 1 above)
Comment (theyueli) [reply src/server/main_service.cc:1488]: fixed.

File: src/server/main_service.cc (hunk 3 above)
Comment (theyueli) [reply src/server/main_service.cc:1164]: > 1. please add a test with move, zinterstore etc (commands that have two hops)

sounds good, I will include all the functional tests in the second PR, which will completes the implementation with invalidation logic.

**Comment by theyueli** [comment]
@dranikpg i addressed all the comments, please take another look.

**Review by dranikpg — APPROVED** [review]

File: src/server/db_slice.cc [hunk 7]
```diff
@@ -1338,4 +1338,19 @@ void DbSlice::ResetUpdateEvents() {
   events_.update = 0;
 }
 
+void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice& keys) {
+  if (conn.IsExpired()) {
```
Comment (dranikpg) [inline src/server/db_slice.cc:1342]: This is impossible, I'd do `DCHECK(!conn.Expired())` (in the next pr)

Comment (theyueli) [reply src/server/db_slice.cc:1342]: > This is impossible, I'd do `DCHECK(!conn.Expired())` (in the next pr)

got it, will do it there.

File: src/server/server_family.cc [hunk 8]
```diff
@@ -1341,6 +1343,30 @@ void ServerFamily::ClientPause(CmdArgList args, ConnectionContext* cntx) {
...
+void ServerFamily::ClientTracking(CmdArgList args, ConnectionContext* cntx) {
+  if (args.size() != 1)
+    return cntx->SendError(kSyntaxErr);
+
+  auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  if (!rb->IsResp3())
+    return cntx->SendError(
+        "Client tracking is currently not supported for RESP2. Please use RESP3.");
+
+  ToUpper(&args[0]);
```
Comment (adiholden) [inline src/server/server_family.cc:1355]: ` CmdArgParser parser(args);
bool is_on = false;
  if (parser.HasNext()) {
    pause_state = parser.ToUpper().Switch("ON", true, "OFF", false);
  }
  if (auto err = parser.Error(); err) {
    return (*cntx)->SendError(err->MakeReply());
  }`

Comment (adiholden) [inline src/server/server_family.cc:1355]: CmdArgParser parser(args);
bool is_on = false;
  if (parser.HasNext()) {
    pause_state = parser.ToUpper().Switch("ON", true, "OFF", false);
  }
  if (auto err = parser.Error(); err) {
    return (*cntx)->SendError(err->MakeReply());
  }

File: src/server/db_slice.h [hunk 9]
```diff
@@ -384,6 +388,16 @@ class DbSlice {
...
   DocDeletionCallback doc_del_cb_;
+
+  struct Hash {
+    size_t operator()(const facade::Connection::WeakRef& c) const {
+      return std::hash<uint32_t>()(c.GetClientId());
+    }
+  };
+
+  // the table that maps keys to the clients that are tracking them.
+  absl::flat_hash_map<std::string, absl::flat_hash_set<facade::Connection::WeakRef, Hash>>
```
Comment (adiholden) [inline src/server/db_slice.h:399]: what is the number of keys in this map? relay spoke about 1 milion keys tracked right?

Comment (adiholden) [reply src/server/db_slice.h:399]: we should add tracking to the memory usage here

Comment (adiholden) [reply src/server/db_slice.h:399]: @theyueli ping on this one

**Pull request merged**